src/views/lighttable.c

```diff
@@ -1183,66 +1183,58 @@ int expose_full_preview(dt_view_t *self, cairo_t *cr, int32_t width, int32_t hei
+                                &stmt, NULL);
+    if(sqlite3_step(stmt) == SQLITE_ROW)
     {
-      /* We need to augment the current main query a bit to fetch the
-       * row we need. */
-      const char *main_query = sqlite3_sql(lib->statements.main_query);
-      stmt_string = g_strdup_printf("SELECT col.imgid AS id, col.rowid FROM (%s) AS s1 %s", main_query,
-                                    filter_criteria);
-
-      DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db), stmt_string, -1, &stmt, NULL);
-      DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, 0);
-      DT_DEBUG_SQLITE3_BIND_INT(stmt, 2, -1);
+      sel_img_count = sqlite3_column_int(stmt, 0);
     }
-    g_free(stmt_string);
-    g_free(filter_criteria);
+    sqlite3_finalize(stmt);
+  }
 
-    if(sqlite3_step(stmt) == SQLITE_ROW)
+  sqlite3_stmt *stmt;
+  gchar *stmt_string = NULL;
+  /* How many images to preload in advance. Three is just a random preference. */
+  int preload_num = 3;
+  stmt_string = g_strdup_printf("SELECT col.imgid AS id, col.rowid FROM memory.collected_images AS col %s WHERE col.rowid %s %d ORDER BY col.rowid %s LIMIT %d",
+                                (sel_img_count <= 1) ?
+                                  /* We want to operate on the currently collected images, so there's no need to match against the selection */
+                                  "" :
+                                  /* Limit the matches to the current selection */
+                                  "INNER JOIN selected_images AS sel ON col.imgid = sel.imgid",
+                                (offset >= 0) ? ">" : "<",
+                                lib->full_preview_rowid,
+                                /* Direction of our navigation -- when showing for the first time, i.e. when offset == 0, assume forward navigation */
+                                (offset >= 0) ? "ASC" : "DESC",
+                                preload_num + 1);
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db), stmt_string, -1, &stmt, NULL);
+
+  /* Walk through the "next" images, activate preload and find out where to go if moving */
+  dt_mipmap_size_t mip = dt_mipmap_cache_get_matching_size(darktable.mipmap_cache, width, height);
+  int first_handled = FALSE;
+  while (sqlite3_step(stmt) == SQLITE_ROW)
+  {
+    /* Check if we're about to move */
+    if(!first_handled && offset)
     {
+      first_handled = TRUE;
+      /* We're moving, so let's update the "next image" bits */
       lib->full_preview_id = sqlite3_column_int(stmt, 0);
       lib->full_preview_rowid = sqlite3_column_int(stmt, 1);
       dt_control_set_mouse_over_id(lib->full_preview_id);
     }
-
-    sqlite3_finalize(stmt);
+    /* Always preload the next few images in the correct direction */
+    dt_mipmap_buffer_t buf;
+    dt_mipmap_cache_get(darktable.mipmap_cache, &buf, sqlite3_column_int(stmt, 0), mip, DT_MIPMAP_PREFETCH, 'r');
```

Comment:
You want to prefetch them backwards as these jobs are added to a stack, so the last one added gets run first.